app/src/main/java/com/pvp/app/service/AutocompleteService.kt

```diff
@@ -100,27 +105,112 @@ class TaskAutocompleteService : Service() {
         }
     }
 
+    private suspend fun checkGoalCompletion(
+        goals: List<Goal>
+    ): List<Goal> {
+        val exercises = healthConnectService
+            .readActivityData(
+                record = ExerciseSessionRecord::class,
+                start = LocalDate
+                    .now()
+                    .minusDays(30)
+                    .atStartOfDay(ZoneId.systemDefault())
+                    .toInstant(),
+                end = Instant.now()
+            )
+            .map {
+                exerciseService.getExerciseInfo(it)
+            }
+
+        return goals.map { goal ->
+            if (goal.completed) return@map goal
+
+            when (goal.steps) {
+                true -> {
+                    goal.progress = getSteps(goal).toDouble()
+                }
+
+                false -> {
+                    goal.progress = getDistance(
+                        exercises,
+                        goal
+                    )
+                }
+            }
+
+            if (goal.progress >= goal.target) {
+                goal.completed = true
```

Comment:
I believe `goal.completed = goal.progress >= goal.target` would be simpler while giving the same result (since we are already working with not completed goals, so if we mark them false again, no problem). Once again, just my opinion. If you prefer this way, leave as it is, it is completely fine.